core/src/main/java/org/apache/iceberg/PartitionSpecParser.java

```diff
@@ -146,8 +150,13 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
       String name = JsonUtil.getString(NAME, element);
       String transform = JsonUtil.getString(TRANSFORM, element);
       int sourceId = JsonUtil.getInt(SOURCE_ID, element);
-
-      builder.add(sourceId, name, transform);
+      // to handle the backward compatibility where partitionFieldId was not part of the partitionSpec schema.
+      if (element.has(FIELD_ID)) {
+        partitionFieldId = JsonUtil.getInt(FIELD_ID, element);
+      } else {
+        partitionFieldId = partitionFieldId + 1;
+      }
+      builder.add(sourceId, partitionFieldId, name, transform);
```

Comment:
It seems odd that `partitionFieldId` is incremented from the last value when missing. If a spec has one missing field ID, then it will be assigned based on the previous field's ID. I don't think this would cause problems because we expect either all fields to have assigned IDs, or no fields to have them.

I'd prefer to keep the logic for those cases separate to make this easier to follow. It isn't a good practice to rely on a hidden assumption that either all fields have ids or none do.